test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/MaxInputBufferLengthTests.cs

```diff
@@ -0,0 +1,233 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.FunctionalTests
+{
+    public class MaxInputBufferLengthTests
+    {
+        private const int _dataLength = 10 * 1024 * 1024;
+
+        public static IEnumerable<object[]> LargeUploadData
+        {
+            get
```

Comment:
I will add a comment for each buffer size explaining why it's worth testing.

I do think the buffer sizes (and whether we expect the client to pause) should be passed as theory data.  Creating a new test for each buffer size would be a lot of extra work.

For the 4 combinations of `sendContentLengthHeader` and `ssl`, it would be reasonable to create 4 tests like:

1. `LargeUploadWithContentLength`
2. `LargeUploadWithoutContentLength`
3. `LargeUploadWithContentLengthOverSsl`
4. `LargeUploadWithoutContentLengthOverSsl`

But I still think passing these as boolean parameters is cleaner.